pkg/controller/utils/conditions.go

```diff
@@ -124,9 +126,36 @@ func SetClusterDeploymentConditionWithChangeCheck(
 			changed = true
 		}
 	}
+	if changed {
+		conditions = sortClusterDeploymentConditions(conditions)
+	}
 	return conditions, changed
 }
 
+func sortClusterDeploymentConditions(conditions []hivev1.ClusterDeploymentCondition) []hivev1.ClusterDeploymentCondition {
+	var unknown []hivev1.ClusterDeploymentCondition
+	var desired []hivev1.ClusterDeploymentCondition
+	var undesired []hivev1.ClusterDeploymentCondition
+
+	for _, condition := range conditions {
+		if condition.Status == corev1.ConditionUnknown {
+			unknown = append(unknown, condition)
+		} else if (IsConditionWithPositivePolarity(condition.Type) && condition.Status == corev1.ConditionFalse) ||
+			(!IsConditionWithPositivePolarity(condition.Type) && condition.Status == corev1.ConditionTrue) {
+			undesired = append(undesired, condition)
+		} else {
+			desired = append(desired, condition)
+		}
+	}
+
+	for _, slice := range [][]hivev1.ClusterDeploymentCondition{unknown, desired, undesired} {
+		sort.SliceStable(slice, func(i, j int) bool {
+			return slice[i].Type < slice[j].Type
+		})
```

Comment:
I believe you can remove sortClusterDeploymentConditions function and the need for three slices to be combined, and instead just use the SliceStable sort where we keep falling back to more specific checks for sorting. In that sort function we can do several checks to determine if I should be sorted before J.

I don't know if my true/falses are right here but hopefully this clarifies the idea:

    if J is Unknown and I isn't, return true
    if I is Unkown and J isn't, return false
    if I is out of polarity state and J isn't, return true
    if J is out of polarity state and I isn't, return false
    lastly do the Type string comparison if none of the above was true

I might be wrong but I'd be surprised if string < otherString is doing what you want? Feels like you need strings.Compare().

Definitely want unit tests on SetClusterDeploymentConditionWithChangeCheck to verify they're being sorted the way you expect.

Reply:
As per SliceStable example [here](https://golang.org/pkg/sort/#SliceStable) string < otherString should work if we are sorting alphabetically.

+1 for unit tests here.